functions/src/auto-importer.ts

```diff
@@ -0,0 +1,263 @@
+import fs from "fs"
+import path from "path"
+import util from "util"
+import crypto from "crypto";
+import * as functions from "firebase-functions";
+import admin, { firestore } from "firebase-admin";
+
+import { S3Client, PutObjectCommand, DeleteObjectCommand } from "@aws-sdk/client-s3";
+
+const parquet = require('parquetjs');
+
+const access = util.promisify(fs.access);
+const unlink = util.promisify(fs.unlink);
+const readFile = util.promisify(fs.readFile);
+
+import { AnswerData, schema, parquetInfo } from "./shared/s3-answers"
+
+/*
+
+  TODO:
+
+  1. Figure out how to get syncSource from the environment
+  2. Figure out how to get AWS config from the environment
+
+*/
+
+/*
+
+HOW THIS WORKS:
+
+1. createSyncDocAfterAnswerWritten runs on every write of an answer.  If the answer changes or is deleted a "sync doc"
+   is created or updated whose name is the run_key for the answer along with a boolean "updated".
+2. monitorSyncDocCount runs as a cron job every few minutes to find all the docs with updated = true.  Each document that is
+   found has a needs_sync field set to the current server time and has updated set to false.
+3. syncToS3AfterSyncDocWritten runs on every write of a sync doc.  If the needs_sync field exists and either the doc has
+   never synced before, or needs_sync > did_sync, it will gather up all the answers for that runKey and post them to
+   s3 as a parquet file. If there are no answers, it will delete the parquet file. It will then set a did_sync timestamp
+4. After the did_sync is set, syncToS3AfterSyncDocWritten will be notified again. If no new data has been written, it will
+   delete the sync doc
```

Comment:
If an answer is changed at the same time that syncToS3AfterSyncDocWritten is running, is it possible for it to delete the document which just had its 'updated' boolean set? 
In other words is it possible that the document fetched by syncToS3AfterSyncDocWritten could be changed before it is deleted?  

Perhaps the delete operation can be run in a transaction that includes the fetching of the document. I'd guess in that case the transaction will be aborted if the document changes. Or the operation to change the document will be blocked until the transaction has completed.